src/libopensc/card-piv.c

```diff
@@ -507,207 +844,1980 @@ static struct sc_card_driver piv_drv = {
+	sc_format_asn1_entry(asn1_piv_cvc_alg_id + 1, NULL, NULL, 1); /* NULL */
+
+	sc_format_asn1_entry(asn1_piv_cvc_dssig    , &asn1_piv_cvc_alg_id, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_dssig + 1, &cvc->signature, &signaturebits, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc_dsobj    , &asn1_piv_cvc_dssig, NULL, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc_pubkey    , &cvc->pubKeyOID, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_pubkey + 1, &cvc->publicPoint, &cvc->publicPointlen, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc_body    , &cvc->cpi, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 1, &cvc->issuerID, &cvc->issuerIDlen, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 2, &cvc->subjectID, &cvc->subjectIDlen, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 3, &asn1_piv_cvc_pubkey, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 4, piv_get_asn1_obj, &roleIDder, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 5, &asn1_piv_cvc_dsobj, NULL, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc, &asn1_piv_cvc_body, NULL, 1);
+
+	r = sc_asn1_decode(card->ctx, asn1_piv_cvc, *buf, *buflen, NULL, NULL) ; /*(const u8 **) &buf_tmp, &len);*/
+	if (r < 0) {
+		piv_clear_cvc_content(cvc);
+		sc_log(card->ctx, "Could not decode card verifiable certificate");
+		LOG_FUNC_RETURN(card->ctx, r);
+	}
+
+	cvc->signaturelen = signaturebits / 8;
+
+	if (roleIDder.len != 1)
+		LOG_TEST_RET(card->ctx, SC_ERROR_SM_AUTHENTICATION_FAILED, "roleID wrong length");
+
+	cvc->roleID = *roleIDder.value;
+
+	/* save body der for verification */
+	buf_tmp = *buf;
+	r = sc_asn1_read_tag(&buf_tmp, *buflen, &cla_out, &tag_out, &taglen);
+	LOG_TEST_RET(card->ctx, r," failed to read tag");
+
+	cvc->bodylen = (roleIDder.value + roleIDder.len) - buf_tmp;
+
+	cvc->body = malloc(cvc->bodylen);
+	if (cvc->body == NULL)
+		return SC_ERROR_OUT_OF_MEMORY;
+	memcpy(cvc->body, buf_tmp, cvc->bodylen);
+
+	/* save to reuse */
+	cvc->der.value = malloc(*buflen);
+	if (cvc->der.value == NULL) {
+		free(cvc->body);
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+	cvc->der.len = *buflen;
+	memcpy(cvc->der.value, *buf, cvc->der.len);
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+static int piv_parse_pairing_code(sc_card_t *card, const char *option)
+{
+	size_t i;
+
+	if (strlen(option) != PIV_PAIRING_CODE_LEN) {
+		sc_log(card->ctx, "pairing code length invalid must be %d", PIV_PAIRING_CODE_LEN);
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	for (i = 0; i < PIV_PAIRING_CODE_LEN; i++) {
+		if (!isdigit(option[i])) {
+			sc_log(card->ctx, "pairing code must be %d decimal digits",PIV_PAIRING_CODE_LEN);
+			return SC_ERROR_INVALID_ARGUMENTS;
+		}
+	}
+	return SC_SUCCESS;
+}
+#endif
+
+static int piv_load_options(sc_card_t *card)
+{
+	int r;
+	piv_private_data_t * priv = PIV_DATA(card);
+	size_t i, j;
+	scconf_block **found_blocks, *block;
+
+	const char *option = NULL;
+#ifdef ENABLE_PIV_SM
+	int piv_pairing_code_found = 0;
+	int piv_use_sm_found = 0;
+#endif
+
+#ifdef ENABLE_PIV_SM
```

Comment:
nit: needless close and open of `ifdef`.

Reply:
> nit: needless close and open of ifdef.

 fixed in 2e9991c792b8ee4c00398af962771fe579b39886